urldownloader.py

```diff
@@ -124,6 +124,136 @@ def __init__(self, config=None) -> None:
+                kangooroo_config["kang-upstream-proxy"]["password"] = url_proxy.password
+        else:
+            kangooroo_config.pop("kang-upstream-proxy", None)
+
+        # create the file that we use to run Kangooroo
+        with tempfile.NamedTemporaryFile(dir=self.working_directory, delete=False, mode="w") as temp_conf:
+            yaml.dump(kangooroo_config, temp_conf)
+
+        # Set up environment variable and commandline arguments for running kangooroo
+        env_variables = {"JAVA_OPTS": f"-Xmx{math.floor(self.service_attributes.docker_config.ram_mb*0.75)}m"}
+        kangooroo_args = [
+            "./bin/kangooroo",
+            # We need no-sandbox to run google chrome in a pod
+            "--no-sandbox",
+            "--conf-file",
+            temp_conf.name,
+            "-mods",
+            "summary,captcha",
+            "--simple-result",
+            "--url",
+            request.task.fileinfo.uri_info.uri,
+        ]
+
+        try:
+            subprocess.run(kangooroo_args, cwd=KANGOOROO_FOLDER, timeout=self.request_timeout, env=env_variables)
+        except subprocess.TimeoutExpired:
+            request.partial()
+            timeout_section = ResultTextSection("Request timed out", parent=request.result)
+            timeout_section.add_line(
+                f"Timeout of {self.request_timeout} seconds was not enough to process the query fully."
+            )
+            return None, None
+
+        url_md5 = hashlib.md5(request.task.fileinfo.uri_info.uri.encode()).hexdigest()
+
+        output_folder = os.path.join(kangooroo_config["output_folder"], url_md5)
+        if not os.path.exists(output_folder):
+            possible_folders = os.listdir(kangooroo_config["output_folder"])
+            if len(possible_folders) == 0:
+                raise Exception(
+                    (
+                        "No Kangooroo output folder found. Kangooroo may have been OOMKilled. "
+                        "Check for memory usage and increase limit as needed."
+                    )
+                )
+            elif len(possible_folders) != 1:
+                raise Exception(
+                    (
+                        "Multiple Kangooroo output folders found. Unknown situation happened, you can try "
+                        "submitting this URL again to see if it would help."
+                    )
+                )
+            else:
+                url_hash_mismatch = ResultTextSection("URL hash mismatch", parent=request.result)
+                url_hash_mismatch.add_line(
+                    (
+                        f"URL '{request.task.fileinfo.uri_info.uri}' ({url_md5}) was requested "
+                        f"but a different URL was fetched ({possible_folders[0]})."
+                    )
+                )
+                output_folder = os.path.join(kangooroo_config["output_folder"], possible_folders[0])
+
+        results_filepath = os.path.join(output_folder, "results.json")
```

Comment:
You do not need results_filepath since you are calculating it on L285. You can return a single element here.